File: src/main/java/net/thedragonteam/thedragonlib/network/MessageHandlerWrapper.java
package net.thedragonteam.thedragonlib.network;

import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public abstract class MessageHandlerWrapper<REQ extends IMessage, REPLY extends IMessage> implements IMessageHandler<REQ, REPLY> {

    @Override
    public REPLY onMessage(REQ message, MessageContext ctx) {

        PacketSyncObject<REQ, REPLY> syncObject = new PacketSyncObject<REQ, REPLY>(message, ctx) {
            @Override
            public void run() {
                reply = handleMessage(message, ctx);
            }
        };

        switch (ctx.side) {
            case CLIENT:
                syncObject.addPacketClient();
                break;
            default:
                syncObject.addPacketServer();
                break;
        }

        return syncObject.reply;
    }

    public abstract REPLY handleMessage(REQ message, MessageContext ctx);

}

File: src/main/java/net/thedragonteam/thedragonlib/network/MessageHandlerWrapper.kt
package net.thedragonteam.thedragonlib.network

import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.Side

abstract class MessageHandlerWrapper<REQ : IMessage, REPLY : IMessage> : IMessageHandler<REQ, REPLY> {

    override fun onMessage(message: REQ, ctx: MessageContext): REPLY {

        val syncObject = object : PacketSyncObject<REQ, REPLY>(message, ctx) {
            override fun run() {
                reply = handleMessage(message, ctx)
            }
        }

        when (ctx.side) {
            Side.CLIENT -> syncObject.addPacketClient()
            else -> syncObject.addPacketServer()
        }

        return syncObject.reply as REPLY
    }

    abstract fun handleMessage(message: REQ, ctx: MessageContext): REPLY?

}